win32ss/user/ntuser/ghost.c

```diff
@@ -62,89 +120,297 @@ BOOL FASTCALL IntIsGhostWindow(PWND Window)
+    if (!pTargetWnd)
     {
-        DPRINT("Not a window\n");
-        return NULL;
+        ERR("ValidateHwndNoErr: %p\n", (void *)hwndTarget);
+        goto Quit;
     }
-    return IntGhostWindowFromHungWindow(pHungWnd);
+
+    // get target info
+    style = pTargetWnd->style;
+    exstyle = pTargetWnd->ExStyle;
+    IntGetWindowRect(pTargetWnd, &rc);
+    if (rc.left >= rc.right || rc.top >= rc.bottom)
+    {
+        ERR("Rect empty: %p\n", (void *)hwndTarget);
+        goto Quit;
+    }
+
+    // don't use scrollbars.
+    style &= ~(WS_HSCROLL | WS_VSCROLL | WS_VISIBLE);
+
+    // initialize CREATESTRUCT
+    RtlZeroMemory(&Cs, sizeof(Cs));
+    Cs.lpCreateParams = hwndTarget;
+    Cs.hInstance = NULL;
+    Cs.hMenu = NULL;
+    Cs.hwndParent = NULL;
+    Cs.cy = rc.bottom - rc.top;
+    Cs.cx = rc.right - rc.left;
+    Cs.y = rc.top;
+    Cs.x = rc.left;
+    Cs.style = style;
+    Cs.lpszName = NULL;
+    Cs.lpszClass = GhostClass.Buffer;
+    Cs.dwExStyle = exstyle;
+
+    // WindowName
+    RtlZeroMemory(&WindowName, sizeof(WindowName));
+
+    // create the ghost
+    pGhostWnd = co_UserCreateWindowEx(&Cs, &GhostClass, &WindowName, NULL);
+    if (!pGhostWnd)
+    {
+        ERR("Failed to create ghost\n");
+        goto Quit;
+    }
+
+    // add to list
+    bRet = IntAddGhostEntry(hwndTarget, pGhostWnd->head.h);
+
+Quit:
+    if (!bRet && pGhostWnd)
+    {
+        // destroy ghost
+        co_UserDestroyWindow(pGhostWnd);
+    }
+
+    // resume hooks
+    if (!NoHooks)
+    {
+        ptiCurrent->TIF_flags &= ~TIF_DISABLEHOOKS;
+        ptiCurrent->pClientInfo->dwTIFlags = ptiCurrent->TIF_flags;
+    }
+
+    return bRet;
 }
 
-HWND FASTCALL IntHungWindowFromGhostWindow(PWND pGhostWnd)
+BOOL FASTCALL
+IntGhostOnDestroy(HWND hwndGhost, HWND hwndTarget)
 {
-    const GHOST_DATA *UserData;
-    HWND hwndTarget;
+    PWND pGhostWnd;
+
+    pGhostWnd = ValidateHwndNoErr(hwndGhost);
+    if (!pGhostWnd)
+    {
+        ERR("Not a window.\n");
+        return FALSE;
+    }
 
     if (!IntIsGhostWindow(pGhostWnd))
     {
-        DPRINT("Not a ghost window\n");
-        return NULL;
+        ERR("Not a Ghost window.\n");
+        return FALSE;
     }
 
-    UserData = (const GHOST_DATA *)pGhostWnd->dwUserData;
-    if (UserData)
+    return IntDeleteGhostEntry(hwndTarget);
+}
+
+/*
+ * GhostThreadMain
+ *
+ * Creates ghost windows and exits when no non-responsive window remains.
+ */
+VOID NTAPI
+UserGhostThreadEntry(VOID)
+{
+    // This thread should handle all ghost windows and exit when no ghost window is needed.
+    BOOL bInitialized = FALSE;
+    MSG msg;
+    HDESK hDesk;
+
+    TRACE("UserGhostThreadEntry: started.\n");
+
+    ASSERT(!UserIsEntered());
+    UserEnterExclusive();
+
+    // attach thread to desktop
+    hDesk = IntGetDesktopObjectHandle(gpdeskInputDesktop);
+    IntSetThreadDesktop(hDesk, FALSE);
+
+    // get a PTHREADINFO
+    gptiGhostThread = PsGetCurrentThreadWin32Thread();
```

Comment:
Can you add a FIXME here that mentions we're missing a thread lock (what Windows calls UserRefObjectCo)? If this thread gets killed in user mode during one of the call-outs, gptiGhostThread will point to an invalid thread. A thread lock would handle this case.

Reply:
OK.